src/v/serde/avro/parser.cc

```diff
@@ -0,0 +1,218 @@
+            auto element_count = decode_element_count();
+            parsed::map ret;
+            if (element_count > 0) {
+                ret.entries.reserve(element_count);
+                for (size_t i = 0; i < element_count; ++i) {
+                    auto key = parse_string();
+                    ret.entries.emplace_back(
+                      std::move(key),
+                      co_await parse_node(node->leafAt(1), state));
+                }
+                // avro always writes 0 as the map footer, first element count
+                // field is skipped if the map is empty
+                _parser.read_varlong();
+            }
+            co_return std::make_unique<parsed::message>(std::move(ret));
+        }
+        case ::avro::AVRO_UNION: {
+            auto [idx, _] = _parser.read_varlong();
+            if (idx < 0 || idx >= static_cast<int64_t>(node->leaves())) {
+                throw std::invalid_argument(fmt::format(
+                  "Invalid union branch decoded, it must be greater than 0 and "
+                  "smaller than number of leaves in union {}. current: {}",
+                  node->leaves(),
+                  idx));
+            }
+
+            auto union_branch = node->leafAt(idx);
+            co_return std::make_unique<parsed::message>(parsed::avro_union{
+              .branch = static_cast<size_t>(idx),
+              .message = co_await parse_node(union_branch, state)});
+        }
+        case ::avro::AVRO_FIXED: {
+            // fixed type is special as it does not have the size included in
+            // the binary representation
+            if (_parser.bytes_left() < node->fixedSize()) {
+                throw std::invalid_argument(fmt::format(
+                  "Error parsing fixed avro type. Expected fixed size: {}, "
+                  "while only {} bytes left in the buffer",
+                  node->fixedSize(),
+                  _parser.bytes_left()));
+            }
+
+            co_return std::make_unique<parsed::message>(
+              _parser.share(node->fixedSize()));
+        }
+        case ::avro::AVRO_SYMBOLIC:
+            // this is a special case used when dealing with nested types, we
+            // need to resolve the original type and parse it
+            co_return co_await parse_node(::avro::resolveSymbol(node), state);
+        case ::avro::AVRO_UNKNOWN:
+            throw std::logic_error{
+              "serde avro parser encountered an unknown type"};
+        }
+    }
+
+    size_t parse_length() {
+        auto [sz, _] = _parser.read_varlong();
+        validate_length(sz);
+        return sz;
+    }
+
+    iobuf parse_string() {
+        auto sz = parse_length();
+        return _parser.share(sz);
+    }
+
+    size_t decode_element_count() {
+        auto [value, _] = _parser.read_varlong();
+        if (value < 0) {
+            _parser.read_varlong();
+            value = -value;
+        }
+        validate_length(value);
```

Comment:
maybe a comment that this is a best effort check against corruption? Because value is element count, and not byte count.